File: solutions/java/go-counting/1/src/main/java/GoCounting.java
// For my Shree DR.MDD

import java.awt.Point;
import java.util.*;

class GoCounting {
    private final char[][] grid;
    private final int totalRows;
    private final int totalCols;

    public GoCounting(String boardString) {
        String[] lines = boardString.split("\n");
        this.totalRows = lines.length;
        this.totalCols = lines[0].length();
        this.grid = new char[totalRows][totalCols];

        for (int i = 0; i < totalRows; i++) {
            this.grid[i] = lines[i].toCharArray();
        }
    }

    public Player getTerritoryOwner(int x, int y) {
        if (!isValidCoordinate(x, y)) throw new IllegalArgumentException("Invalid coordinate");
        if (grid[y][x] != ' ') return Player.NONE;
        return exploreTerritory(x, y).owner;
    }

    public Set<Point> getTerritory(int x, int y) {
        if (!isValidCoordinate(x, y)) throw new IllegalArgumentException("Invalid coordinate");
        if (grid[y][x] != ' ') return Collections.emptySet();
        return exploreTerritory(x, y).territory;
    }

    public Map<Player, Set<Point>> getTerritories() {
        Map<Player, Set<Point>> allTerritories = new HashMap<>();
        allTerritories.put(Player.BLACK, new HashSet<>());
        allTerritories.put(Player.WHITE, new HashSet<>());
        allTerritories.put(Player.NONE, new HashSet<>());

        boolean[][] visited = new boolean[totalRows][totalCols];

        for (int row = 0; row < totalRows; row++) {
            for (int col = 0; col < totalCols; col++) {
                if (grid[row][col] == ' ' && !visited[row][col]) {
                    TerritoryResult result = exploreTerritory(col, row);
                    allTerritories.get(result.owner).addAll(result.territory);
                    for (Point p : result.territory) visited[p.y][p.x] = true;
                }
            }
        }

        return allTerritories;
    }

    private boolean isValidCoordinate(int x, int y) {
        return x >= 0 && x < totalCols && y >= 0 && y < totalRows;
    }

    private TerritoryResult exploreTerritory(int startX, int startY) {
        Set<Point> territorySet = new HashSet<>();
        Set<Character> bordering = new HashSet<>();
        Queue<Point> queue = new LinkedList<>();
        queue.add(new Point(startX, startY));
        boolean[][] visited = new boolean[totalRows][totalCols];
        visited[startY][startX] = true;

        int[][] directions = {{0,1}, {1,0}, {0,-1}, {-1,0}};

        while (!queue.isEmpty()) {
            Point current = queue.poll();
            territorySet.add(current);

            for (int[] dir : directions) {
                int nx = current.x + dir[0];
                int ny = current.y + dir[1];

                if (isValidCoordinate(nx, ny)) {
                    char cell = grid[ny][nx];
                    if (cell == ' ' && !visited[ny][nx]) {
                        queue.add(new Point(nx, ny));
                        visited[ny][nx] = true;
                    } else if (cell == 'B' || cell == 'W') {
                        bordering.add(cell);
                    }
                }
            }
        }

        Player owner;
        if (bordering.isEmpty() || bordering.size() > 1) owner = Player.NONE;
        else owner = bordering.contains('B') ? Player.BLACK : Player.WHITE;

        return new TerritoryResult(owner, territorySet);
    }

    private static class TerritoryResult {
        final Player owner;
        final Set<Point> territory;

        TerritoryResult(Player owner, Set<Point> territory) {
            this.owner = owner;
            this.territory = territory;
        }
    }
}

File: solutions/javascript/go-counting/1/go-counting.js
// For my Shree DR.MDD

export class GoCounting {
  constructor(boardMatrix) {
    this.blackTerritory = [];
    this.whiteTerritory = [];
    this.noTerritory = [];
    this.stones = [];

    if (!boardMatrix.toString().includes("B") && !boardMatrix.toString().includes("W")) {
      boardMatrix.forEach((row, rowIndex) => {
        for (let colIndex = 0; colIndex < row.length; colIndex++) {
          this.noTerritory.push([colIndex, rowIndex]);
        }
      });
    }

    const markNoTerritory = (x, y) => {
      boardMatrix[y] = boardMatrix[y].slice(0, x) + "x" + boardMatrix[y].slice(x + 1);
      this.noTerritory.push([x, y]);
      if (x > 0 && boardMatrix[y][x - 1] === " ") markNoTerritory(x - 1, y);
      if (x < boardMatrix[y].length - 1 && boardMatrix[y][x + 1] === " ") markNoTerritory(x + 1, y);
      if (y > 0 && boardMatrix[y - 1][x] === " ") markNoTerritory(x, y - 1);
      if (y < boardMatrix.length - 1 && boardMatrix[y + 1][x] === " ") markNoTerritory(x, y + 1);
    };

    const markBlackTerritory = (x, y) => {
      this.blackTerritory.push([x, y]);
      boardMatrix[y] = boardMatrix[y].slice(0, x) + "b" + boardMatrix[y].slice(x + 1);
      if (x > 0 && boardMatrix[y][x - 1] === " ") markBlackTerritory(x - 1, y);
      if (x < boardMatrix[y].length - 1 && boardMatrix[y][x + 1] === " ") markBlackTerritory(x + 1, y);
      if (y > 0 && boardMatrix[y - 1][x] === " ") markBlackTerritory(x, y - 1);
      if (y < boardMatrix.length - 1 && boardMatrix[y + 1][x] === " ") markBlackTerritory(x, y + 1);
    };

    const markWhiteTerritory = (x, y) => {
      this.whiteTerritory.push([x, y]);
      boardMatrix[y] = boardMatrix[y].slice(0, x) + "w" + boardMatrix[y].slice(x + 1);
      if (x > 0 && boardMatrix[y][x - 1] === " ") markWhiteTerritory(x - 1, y);
      if (x < boardMatrix[y].length - 1 && boardMatrix[y][x + 1] === " ") markWhiteTerritory(x + 1, y);
      if (y > 0 && boardMatrix[y - 1][x] === " ") markWhiteTerritory(x, y - 1);
      if (y < boardMatrix.length - 1 && boardMatrix[y + 1][x] === " ") markWhiteTerritory(x, y + 1);
    };

    for (let y = 0; y < boardMatrix.length; y++) {
      for (let x = 0; x < boardMatrix[y].length; x++) {
        let adjacentColor = "";
        const cell = boardMatrix[y][x];
        if (["x", "b", "w"].includes(cell)) continue;
        if (cell === "B" || cell === "W") {
          this.stones.push([x, y]);
          continue;
        }

        if (x > 0) {
          if (boardMatrix[y][x - 1] !== " " && boardMatrix[y][x - 1] !== "x") adjacentColor = boardMatrix[y][x - 1];
          else if (boardMatrix[y][x - 1] === "x") {
            markNoTerritory(x, y);
            continue;
          }
        }

        if (x < boardMatrix[y].length - 1) {
          if (!adjacentColor && boardMatrix[y][x + 1] !== " " && boardMatrix[y][x + 1] !== "x") adjacentColor = boardMatrix[y][x + 1];
          else if (boardMatrix[y][x + 1] === "x") {
            markNoTerritory(x, y);
            continue;
          } else if (adjacentColor && adjacentColor !== boardMatrix[y][x + 1] && boardMatrix[y][x + 1] !== " ") {
            markNoTerritory(x, y);
            continue;
          } else if (boardMatrix[y][x + 1] === " " && this.noTerritory.some(([cx, cy]) => cx === x + 1 && cy === y)) {
            markNoTerritory(x, y);
            continue;
          }
        }

        if (y > 0) {
          if (adjacentColor !== boardMatrix[y - 1][x] && boardMatrix[y - 1][x] !== " ") {
            markNoTerritory(x, y);
            continue;
          } else if (boardMatrix[y - 1][x] === " " && this.noTerritory.some(([cx, cy]) => cx === x && cy === y - 1)) {
            markNoTerritory(x, y);
            continue;
          }
        }

        if (y < boardMatrix.length - 1) {
          if (adjacentColor !== boardMatrix[y + 1][x] && boardMatrix[y + 1][x] !== " ") {
            markNoTerritory(x, y);
            continue;
          } else if (boardMatrix[y + 1][x] === " " && this.noTerritory.some(([cx, cy]) => cx === x && cy === y + 1)) {
            markNoTerritory(x, y);
            continue;
          }
        }

        switch (adjacentColor) {
          case "W":
            markWhiteTerritory(x, y);
            break;
          case "B":
            markBlackTerritory(x, y);
            break;
        }
      }
    }

    this.board = boardMatrix;
  }

  getTerritory(x, y) {
    if (x < 0 || x >= this.board[0].length || y < 0 || y >= this.board.length) {
      return { error: "Invalid coordinate" };
    }

    const output = { owner: "", territory: [] };

    const floodFill = (marker, x, y) => {
      output.territory.push([x, y]);
      this.board[y] = this.board[y].slice(0, x) + "o" + this.board[y].slice(x + 1);
      if (x > 0 && this.board[y][x - 1] === marker) floodFill(marker, x - 1, y);
      if (x < this.board[y].length - 1 && this.board[y][x + 1] === marker) floodFill(marker, x + 1, y);
      if (y > 0 && this.board[y - 1][x] === marker) floodFill(marker, x, y - 1);
      if (y < this.board.length - 1 && this.board[y + 1][x] === marker) floodFill(marker, x, y + 1);
    };

    const cell = this.board[y][x];
    if (cell === "B" || cell === "W" || cell === "x") output.owner = cell === "x" ? "NONE" : "NONE";
    else if (cell === "b") output.owner = "BLACK";
    else if (cell === "w") output.owner = "WHITE";

    if (cell !== "B" && cell !== "W") floodFill(cell, x, y);
    output.territory.sort();

    return output;
  }

  getTerritories() {
    return {
      territoryBlack: this.blackTerritory,
      territoryWhite: this.whiteTerritory,
      territoryNone: this.noTerritory,
    };
  }
}
